### PR TITLE
Feature EventModal

### PR DESCRIPTION
**Pull request by maxeckel**
Added an "Add Event" Form/Modal, which opens when the "Add Event" Button on a Day is clicked.

This also adds an event to the calendar to react to newly added events.

### CODE REVIEW

**Comment by EazyServer** [comment]
Master has now diverged and won't let me merge!

**Comment by maxeckel** [comment]
Conflicts are resovled.

**Comment by EazyServer** [comment]
Unfortunately it is not! Git conflict highlights are showing on your source code!
screen shot 2017-06-10 at 18 29 56

**Comment by maxeckel** [comment]
What the hell... 

Now EVERY conflict and marker is gone for good, i have to review more carefully after such a merge.

**Comment by EazyServer** [comment]
Great work Max!

The functionality is working great. Just couple of things:

- Can you add background shadow when Add event modal pops up?
- Can you add translation to "Add Event" and "Cancel" and colors dropdown?

screen shot 2017-06-10 at 19 23 22

Cheers

File: src/Components/EventModal.vue
Comment (EazyServer) [inline src/Components/EventModal.vue]: I don't this is necessary! you can use text directly. this applies to `cancelText` as well

File: src/Components/EventModal.vue
Comment (EazyServer) [inline src/Components/EventModal.vue]: Can you please add `cancel` & `update:show` events to [actions.js](https://github.com/EazyServer/Vue-Bootstrap-Calendar/blob/master/src/actions.js) 

It would also be great if you name the constants to 

- CANCEL_ADD_EVENT_FORM
- SHOW_ADD_EVENT_FORM

**Comment by maxeckel** [comment]
Yeah no problem, add that right away

File: src/Components/EventModal.vue
Comment (EazyServer) [inline src/Components/EventModal.vue]: You can use text directly here i.e. `$t('generic.cancel')`

File: src/locale/ar.js
Comment (EazyServer) [inline src/locale/ar.js]: event_title = '**عنوان المناسبة**'
event_color = '**لون المناسبة**'

Can you please add the Arabic translation, thanks

**Comment by maxeckel** [comment]
Both added ;)

**Comment by maxeckel** [comment]
I also added an Event Description (this will be good for the Event Details pull request, with which i wait for this to be merged to avoid the troublesome merge conflicts)

**Comment by EazyServer** [comment]
Hi Max,

Can you please review the code and amend as per the comments.

Cheers

File: src/Components/EventModal.vue
Comment (EazyServer) [inline src/Components/EventModal.vue]: Can you please rename the method to `saveEvent`

**Comment by maxeckel** [comment]
Alright, edited the commit message. Have looked through everything and couldn't find any error.
Tested the whole Modal and event creation process again, and everything works as it should.

**Comment by EazyServer** [comment]
There are no errors in your code. everything is working great. 

I just need to unify coding standards in the code. 
So my comments are cosmetics rather than a surgical, if you know what I mean?

**Comment by maxeckel** [comment]
Ok, i've tried my best to format everything as you did (i hope i got you right and that's what you meant :D ).

Maybe you could right a little bit down, how you want the code to be formatted.

**Comment by EazyServer** [comment]
Can you see this on this wall?

screen shot 2017-06-11 at 12 44 58

**Comment by maxeckel** [comment]
AHHH you commented right in the Code. Sorry didn't see that (i mostly work with gitlab which inform me of inline comments in code), i'll look through that in the future.

Edit: I actually can't see those comments.

**Comment by maxeckel** [comment]
I can't see those inline comments (maybe they got lost because of the last squash).

Since i can't answer them there, i will do here.

For the "update:show" event i can't use the constants because this has to be named like this for vue to enable variable syncing. If i move this to action.js i would break the convention there because it would still need to be:

`export const SHOW_ADD_EVENT_FORM = "update:show";`

instead of

`export const SHOW_ADD_EVENT_FORM = "SHOW_ADD_EVENT_FORM";`

And for the Arabic translations it would be great if you could write those down again, including the new added ones for example the colors.

**Comment by EazyServer** [comment]
`export const SHOW_ADD_EVENT_FORM = "update:show";` is completely fine.

```
'event_title' = 'عنوان المناسبة'
'event_color': "لون المناسبة",
'event_description': "وصف المناسبة",
'cancel': "الغاء",
'color_grey': "رمادي",
'color_blue': "أزرق",
'color_lightblue': "سمائي",
'color_green': "أخضر",
'color_orange': "برتقالي",
'color_red': "أحمر",
```

**Comment by maxeckel** [comment]
Alright, added the translation and integrated the suggested changes.

I'm really sorry for the inconvenience.

File: src/App.vue
Comment (EazyServer) [inline src/App.vue]: `@eventAdded="eventAdded"` is not needed. Please remove

File: src/App.vue
Comment (EazyServer) [inline src/App.vue]: `methods` section is not needed for the purpose you using it for. Please remove

File: src/App.vue
Comment (EazyServer) [inline src/App.vue]: Instead of the two omissions above please use the following code:

`this.$root.$on(EVENT_ADDED, function(payload) {
                me.events.push(payload);
            });`

File: src/Calendar.vue
Comment (EazyServer) [inline src/Calendar.vue]: the following code is redundant! Please remove 

`this.$root.$on(EVENT_ADDED, function(payload) {
 me.$emit('eventAdded', payload);
 });`

File: src/Components/Day.vue
Comment (EazyServer) [inline src/Components/Day.vue]: Please update the EventModal with the following:

`<EventModal
                    :show.sync="showAddEvent"
                    :day="day">
            </EventModal>`

**Comment by EazyServer** [comment]
Thanks for that.

I have refactored your code a little to make it easier to follow the logic. please see my comments on the code.

**Review by EazyServer — APPROVED** [review]

**Comment by maxeckel** [comment]
Sadly the comments have just been removed.

About the @eventAdded event. This is for end user convenience, if you know the exact implementation under the hood, yeah then you know about

`this.$root.on(EVENT_ADDED. function(payload) {});`

Also it is easier for the user to just add the

`@eventAdded($method)`

Actually i just used the $root emit to avoid passing up the event through every component until Calendar to make the "shortcut" available to users.

**Comment by EazyServer** [comment]
Hi Max,

I think using  `@eventAdded` in the way above is not best practice!

Usually `@customeEvent` is used when child communicates with parent! But in this case we have a sub-child `Day` emitting custom event to `Calendar` component!
 
The practise I am using in this package, if data flow is child to parent I can use `@customEvent`. Otherwise, I use `$on` hook inside `mounted()`. The reason for that is easy to understand code.

Cheers

**Comment by maxeckel** [comment]
Yes totally agree with that. Usally this is the way i think too, but in this case where a package is built used by others it's easier to understand or used by simply using the @ shortcut, especially because they don't have the consts of action.js per default and will also have to include them or using the string representation.

**Comment by EazyServer** [comment]
That's a very good point.

Let's do it the way you explained but can we re-name the event to be more package specific? I think `eventAdded` is very generic!

I would suggest using `VBC_EVENT_ADDED` as external event name, and re-name `eventAdded()` method to `handleEventAdded()`

 on `Day.vue` file can you please change:
`<EventModal title="Test" :show.sync="showAddEvent" :day="day">
                Testing
            </EventModal>`

To

`<EventModal 
:show.sync="showAddEvent" 
:day="day">
</EventModal>`

**Pull request closed without merging**
